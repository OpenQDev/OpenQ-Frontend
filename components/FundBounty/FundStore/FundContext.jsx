import React from 'react';
const FundContext = React.createContext();
export default FundContext;
