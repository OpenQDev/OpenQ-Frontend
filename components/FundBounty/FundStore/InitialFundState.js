const zeroAddressMetadata = {
  name: 'Matic',
  address: '0x0000000000000000000000000000000000000000',
  symbol: 'MATIC',
  decimals: 18,
  chainId: 80001,
  path: '/crypto-logos/MATIC.svg',
};

// Reminder, this isn't all the props fundState has, just the initial ones.
const InitialFundState = {
  showApproveTransferModal: false,
  volume: '',
  token: zeroAddressMetadata,
  nftTier: '',
  allowance: '',
  successMessage: '',
  depositPeriodDays: 30,
  error: '',
  transactionHash: '',
  approveTransferState: 'RESTING',
};

export default InitialFundState;
