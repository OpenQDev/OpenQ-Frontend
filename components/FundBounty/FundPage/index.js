// Third party
import React, { useState, useContext, useEffect } from 'react';
import { ethers } from 'ethers';

// Custom
import useWeb3 from '../../../hooks/useWeb3';
import TokenFundBox from '../../TokenSelection/TokenFundBox';
import FundContext from '../FundStore/FundContext';
import StoreContext from '../../../store/Store/StoreContext';
import ToolTipNew from '../../Utils/ToolTipNew';
import BountyClosed from '../../BountyClosed';
import ApproveFundModal from '../ApproveFundModal';
import OrgDetails from '../../User/InvoicingDetailsTab/OrgDetails';
import { RESTING, CONFIRM, APPROVING, TRANSFERRING, APPROVE } from '../FundStore/ApproveFundState';
import useIsOnCorrectNetwork from '../../../hooks/useIsOnCorrectNetwork';
import NFTFundModal from '../../TokenSelection/NFTFundModal';
import DepositPeriod from '../../TokenSelection/DepositPeriod';
import ConnectButton from '../../WalletConnect/ConnectButton';
import { ChevronDownIcon, ChevronUpIcon } from '@primer/octicons-react';
import { valueToDisplay, listWordsWithAnd } from '../../../services/utils/lib';
import useFundBountyMethod from '../hooks/useFundBountyMethod';
import TokenContext from '../../TokenSelection/TokenStore/TokenContext';
import { shortenAddress, getBountyTypeName } from '../../../services/utils/lib';
import { useRouter } from 'next/router';

const FundPage = () => {
  const [fundState, fundDispatch] = useContext(FundContext);
  const { allowance, depositPeriodDays, bounty, approveTransferState, setInternalMenu } = fundState;
  const web3 = useWeb3();
  const query = useRouter().query;
  const minter = bounty.issuer.id === web3?.account?.toLowerCase();
  const canCrowdFund = !(bounty.invoiceRequired || query.invoiceable) && !bounty.kycRequired;
  const canFund = canCrowdFund || minter;
  const [volume, setVolume] = useState('');
  const [tokenSelectState] = useContext(TokenContext);
  const { token } = tokenSelectState;
  const [isOnCorrectNetwork] = useIsOnCorrectNetwork();

  const [appState] = useContext(StoreContext);
  const fundBountyMethod = useFundBountyMethod();
  const { accountData, utils } = appState;
  const accountKeys = bounty.invoiceRequired
    ? ['company', 'city', 'country', 'streetAddress', 'province', 'invoicingEmail']
    : ['invoicingEmail'];
  const mustChangePayoutFirst =
    (bounty?.bountyType == '1' && bounty?.payoutTokenVolume > 0) ||
    (bounty?.bountyType == '3' && bounty?.payoutSchedule?.length > 0);
  const neededAccountData = accountKeys.filter((key) => {
    return !accountData[key];
  });
  const [, tokenDispatch] = useContext(TokenContext);
  useEffect(() => {
    const depositTokenAddress = bounty?.deposits[0]?.tokenAddress;
    const payoutTokenAddress = bounty?.payoutTokenAddress;
    if ((bounty?.bountyType == '1' || bounty?.bountyType == '3') && bounty?.deposits?.length > 0) {
      const tokenAddressDispatch = {
        type: 'SET_TOKEN',
        payload: {
          ...appState.tokenClient.getToken(depositTokenAddress),
          address: depositTokenAddress,
        },
      };
      tokenDispatch(tokenAddressDispatch);
      fundDispatch(tokenAddressDispatch);
    } else if (
      (bounty?.bountyType == '1' && bounty?.payoutTokenVolume > 0) ||
      (bounty?.bountyType == '3' && bounty?.payoutSchedule?.length > 0)
    ) {
      const tokenAddressDispatch = {
        type: 'SET_TOKEN',
        payload: {
          ...appState.tokenClient.getToken(payoutTokenAddress),
          address: payoutTokenAddress,
        },
      };
      tokenDispatch(tokenAddressDispatch);
      fundDispatch(tokenAddressDispatch);
    }
  }, [bounty]);
  useEffect(() => {
    if (!((bounty?.bountyType == '1' || bounty?.bountyType == '3') && bounty?.deposits?.length > 0)) {
      const tokenDispatch = {
        type: 'SET_TOKEN',
        payload: token,
      };
      fundDispatch(tokenDispatch);
    }
  }, [token]);

  useEffect(() => {
    const volumeDispatch = {
      type: 'SET_VOLUME',
      payload: volume,
    };
    fundDispatch(volumeDispatch);
  }, [volume]);
  const hasInvoicingInfo =
    neededAccountData.length === 0 || (!bounty.invoiceRequired && !bounty.supportingDocumentsRequired);

  // Context
  const { openQClient } = appState;
  const { account, library } = web3;

  const disabledFundButton =
    approveTransferState == CONFIRM ||
    approveTransferState == APPROVING ||
    approveTransferState == TRANSFERRING ||
    bounty.status == '1' ||
    parseFloat(volume) <= 0.00000001 ||
    parseFloat(volume) > 1000000 ||
    volume == '' ||
    !(parseInt(depositPeriodDays) > 0) ||
    !hasInvoicingInfo;

  // Methods
  const hasEnoughAllowance = async () => {
    const allowanceBigNumber = await openQClient.allowance(library, account, token.address, bounty.bountyAddress);
    const volumeInWei = volume * 10 ** token.decimals;
    const bigNumberVolumeInWei = ethers.BigNumber.from(volumeInWei.toLocaleString('fullwide', { useGrouping: false }));
    const allowanceValue =
      allowanceBigNumber > 0
        ? token.address != ethers.constants.AddressZero && allowanceBigNumber?.gte(bigNumberVolumeInWei)
        : 0;
    const allowanceDispatch = {
      type: 'SET_ALLOWANCE',
      payload: allowanceValue,
    };
    fundDispatch(allowanceDispatch);
  };

  const openFund = async () => {
    await hasEnoughAllowance();
    let dispatchPayload;
    if (allowance || token.address == ethers.constants.AddressZero) {
      dispatchPayload = {
        type: 'SET_APPROVE_TRANSFER_STATE',
        payload: CONFIRM,
      };
    } else {
      dispatchPayload = {
        type: 'SET_APPROVE_TRANSFER_STATE',
        payload: APPROVE,
      };
    }
    fundDispatch(dispatchPayload);
  };

  const fundBounty = async () => {
    fundBountyMethod();
  };

  function onVolumeChange(volume) {
    utils.updateVolume(volume, setVolume);
  }

  // Render
  return (
    <>
      {bounty.status == '1' ? (
        <>
          <BountyClosed bounty={bounty} />
        </>
      ) : (
        <div className='flex-1 pt-4 pb-8 w-full max-w-[1200px] justify-center'>
          {canFund ? (
            <div className='flex flex-col w-full space-y-5 pb-8 items-center md:border rounded-sm border-gray-700'>
              <div className='flex text-3xl w-full text-primary justify-center px-16 py-4 md:bg-[#161b22] md:border-b border-gray-700 rounded-t-sm'>
                Escrow Funds in {getBountyTypeName(bounty.bountyType)} Contract
              </div>
              <div className='flex flex-col space-y-5 w-5/6 pt-2'>
                <div className='flex w-full gap-4'>
                  <>
                    <TokenFundBox
                      onVolumeChange={onVolumeChange}
                      token={token}
                      volume={volume}
                      bounty={bounty}
                      mustChangePayoutFirst={mustChangePayoutFirst}
                      setInternalMenu={setInternalMenu}
                    />
                    <NFTFundModal />
                  </>
                </div>
                <div className='flex gap-4'>
                  <div className='w-full'>
                    <DepositPeriod />

                    <ConnectButton needsGithub={false} nav={false} tooltipAction={'fund this contract!'} />
                    {account && isOnCorrectNetwork && (
                      <ToolTipNew
                        relativePosition={'left-0'}
                        outerStyles={'-top-1 '}
                        groupStyles={'w-min'}
                        innerStyles={'sm:w-40 md:w-60 whitespace-normal'}
                        hideToolTip={!disabledFundButton}
                        toolTipText={
                          !(depositPeriodDays > 0)
                            ? "Please indicate how many days you'd like to fund your contract for."
                            : !hasInvoicingInfo
                            ? 'This bounty requires invoicing information.'
                            : "Please indicate the volume you'd like to fund with. Must be between 0.0000001 and 1,000,000."
                        }
                      >
                        <button
                          className={`text-center px-8 w-min items-center py-0.5  ${
                            disabledFundButton
                              ? 'btn-default w-full cursor-not-allowed'
                              : 'btn-primary bg-green cursor-pointer'
                          } py-1.5`}
                          disabled={disabledFundButton}
                          type='button'
                          onClick={openFund}
                        >
                          <div className='text-center whitespace-nowrap w-full'>Fund</div>
                        </button>
                      </ToolTipNew>
                    )}
                  </div>
                </div>
              </div>

              {(bounty.invoiceRequired || bounty.supportingDocumentsRequired) && (
                <>
                  {bounty.invoiceRequired && neededAccountData.length !== 0 ? (
                    <div className='w-5/6'>
                      Invoicing data required for this bounty, you are missing values for the{' '}
                      {listWordsWithAnd(
                        neededAccountData.map((elem) => {
                          return valueToDisplay(elem);
                        })
                      )}{' '}
                      fields.
                    </div>
                  ) : (
                    <></>
                  )}
                  <details className='w-5/6 group' open={!hasInvoicingInfo}>
                    <summary className='list-none text-2xl text-muted fill-muted cursor-pointer'>
                      {!bounty.invoiceRequired ? 'Contact Information' : 'Invoicing data'}
                      <span className='group-open:hidden'>
                        <ChevronDownIcon size='24px' />
                      </span>
                      <span className='hidden group-open:inline'>
                        <ChevronUpIcon size='24px' />
                      </span>
                    </summary>
                    <OrgDetails emailOnly={!bounty.invoiceRequired && bounty.supportingDocumentsRequired} slim={true} />
                  </details>
                  {!bounty.invoiceRequired && bounty.supportingDocumentsRequired && (
                    <div className='w-5/6 note'>
                      We will send the completed W8/W9 form with a link to accept to this email. Please note that only
                      your ethereum account is able to accept this. If someone else reviews the form you will need to
                      confirm it.
                    </div>
                  )}
                </>
              )}
            </div>
          ) : (
            <div className='p-8 sm:p-0'>
              Sorry crowdfunding isn't available for invoiceable bounties such as this one, please connect the account (
              {shortenAddress(bounty.issuer.id)}) that minted this bounty to fund it.
            </div>
          )}

          {approveTransferState !== RESTING && <ApproveFundModal confirmMethod={fundBounty} />}
        </div>
      )}
    </>
  );
};

export default FundPage;
