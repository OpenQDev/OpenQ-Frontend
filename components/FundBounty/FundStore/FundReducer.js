import { ethers } from 'ethers';
import InitialFundState from './InitialFundState';

const FundReducer = (state, action) => {
  switch (action.type) {
    case 'SET_INVOICEABLE': {
      return {
        ...state,
        invoiceable: action.payload,
      };
    }
    case 'SET_TOKEN': {
      return {
        ...state,
        token: { ...action.payload, address: ethers.utils.getAddress(action.payload.address) },
      };
    }
    case 'SET_VOLUME': {
      return {
        ...state,
        volume: action.payload,
      };
    }
    case 'SET_NFT_TIER': {
      return {
        ...state,
        nftTier: action.payload,
      };
    }
    case 'SET_ALLOWANCE': {
      return {
        ...state,
        allowance: action.payload,
      };
    }
    case 'SET_DEPOSIT_PERIOD_DAYS': {
      return {
        ...state,
        depositPeriodDays: action.payload,
      };
    }
    case 'SET_ERROR': {
      return {
        ...state,
        error: action.payload,
        approveTransferState: 'ERROR',
      };
    }
    case 'SET_APPROVING': {
      return {
        ...state,
        approveTransferState: 'APPROVING',
        approveTransferModal: true,
      };
    }
    case 'SET_TRANSACTION_HASH': {
      return {
        ...state,
        transactionHash: action.payload,
      };
    }
    case 'SET_NFT': {
      return {
        ...state,
        pickedNft: action.payload,
      };
    }
    case 'SET_APPROVE_TRANSFER_STATE': {
      return {
        ...state,
        approveTransferState: action.payload,
      };
    }
    case 'RESET_STATE': {
      return InitialFundState;
    }
    default:
      return state;
  }
};

export default FundReducer;
