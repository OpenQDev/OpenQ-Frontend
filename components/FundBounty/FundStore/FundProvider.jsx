import React, { useReducer } from 'react';
import FundReducer from './FundReducer';
import FundContext from './FundContext';
import InitialState from './InitialFundState';

const FundProvider = ({ children, bounty, refreshBounty, setInternalMenu }) => {
  const [state, dispatch] = useReducer(FundReducer, { bounty, refreshBounty, setInternalMenu, ...InitialState });

  return <FundContext.Provider value={[state, dispatch]}>{children}</FundContext.Provider>;
};

export default FundProvider;
