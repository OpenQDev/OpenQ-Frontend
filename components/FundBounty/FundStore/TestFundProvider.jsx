import React, { useReducer } from 'react';
import FundReducer from './FundReducer';
import FundContext from './FundContext';
import InitialState from './InitialFundState';

const TestFundProvider = ({ children, bounty, refreshBounty, StoreProps }) => {
  const [state, dispatch] = useReducer(FundReducer, { bounty, refreshBounty, ...InitialState, ...StoreProps });

  return <FundContext.Provider value={[state, dispatch]}>{children}</FundContext.Provider>;
};

export default TestFundProvider;
