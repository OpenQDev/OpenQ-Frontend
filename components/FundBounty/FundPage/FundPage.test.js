/**
 * @jest-environment jsdom
 */
import React from 'react';
import { render, screen } from '../../../test-utils';
import FundPage from '.';
import userEvent from '@testing-library/user-event';
import InitialState from '../../../store/Store/InitialState';
import FundContext from '../../FundBounty/FundStore/FundContext';
import FundProvider from '../../FundBounty/FundStore/FundProvider';
import Constants from '../../../test-utils/constant';
import InitialFundState from '../../FundBounty/FundStore/InitialFundState';
import { CONFIRM } from '../../FundBounty/FundStore/ApproveFundState';
import nextRouter from 'next/router';

describe('FundPage', () => {
  const bounty = Constants.bounty;
  const splitBounty = Constants.bounty1;

  const refreshBounty = () => {
    return null;
  };
  beforeEach(() => {
    const observe = jest.fn();
    const disconnect = jest.fn();

    window.IntersectionObserver = jest.fn(() => ({
      observe,
      disconnect,
    }));
    nextRouter.useRouter = jest.fn();
    nextRouter.useRouter.mockImplementation(() => ({
      query: { type: null },

      prefetch: jest.fn(() => {
        return { catch: jest.fn };
      }),
    }));
    InitialState.openQClient.reset();
    InitialState.shouldSleep = 200;
  });

  describe('Split Price Bounties', () => {
    it('should lock token choice to deposit token when there is 1+ deposit', async () => {
      // ARRANGE
      const otherTokenDepositSplitBounty = { ...splitBounty, deposits: [Constants.deposit3] };
      const user = userEvent.setup();
      const { asFragment } = render(
        <FundProvider bounty={otherTokenDepositSplitBounty} refreshBounty={refreshBounty}>
          <FundPage bounty={otherTokenDepositSplitBounty} />
        </FundProvider>
      ); // ACT / ASSERT
      expect(asFragment()).toMatchSnapshot();

      const input = screen.getByLabelText('amount');
      await user.type(input, '0.30sdf');
      expect(screen.getAllByText(/link/i)[0]).toBeInTheDocument;
      const selectToken = screen.getByRole('button', { name: 'select token' });
      expect(selectToken).toBeDisabled();
      expect(selectToken).toMatchSnapshot('Snapshot: button style changes to not appear as button');
      const button = screen.getByRole('button', { name: /Fund/i });
      await user.click(button);
    });
    it('should show modal to update reward split when no deposits but payoutTokenVolume > 0', async () => {
      // ARRANGE
      const noDepositSplitBounty = {
        ...splitBounty,
        deposits: [],
        payoutTokenAddress: '0x5fbdb2315678afecb367f032d93f642f64180aa3',
      };
      const user = userEvent.setup();
      const setInternalMenu = jest.fn();
      render(
        <FundProvider bounty={noDepositSplitBounty} refreshBounty={refreshBounty} setInternalMenu={setInternalMenu}>
          <FundPage bounty={noDepositSplitBounty} />
        </FundProvider>
      ); // ACT / ASSERT

      const matic = screen.queryByText(/matic/i);
      expect(matic).not.toBeInTheDocument();
      const input = screen.getByLabelText('amount');
      await user.type(input, '0.30sdf');
      expect(screen.getAllByText(/link/i)[0]).toBeInTheDocument;
      const selectToken = screen.getByRole('button', { name: 'select token' });
      await user.click(selectToken);
      expect(screen.getAllByText(/Update your reward token first!/i)[0]).toBeInTheDocument;
      const button = screen.getByRole('button', { name: /Update Token/i });
      await user.click(button);
      expect(setInternalMenu).toHaveBeenCalledWith('Admin');
    });
  });

  describe('Fixed Bounties', () => {
    it('should render the heading', async () => {
      // ARRANGE
      const user = userEvent.setup();
      render(
        <FundProvider bounty={bounty} refreshBounty={refreshBounty}>
          <FundPage bounty={bounty} />
        </FundProvider>
      );

      // ACT
      const heading = screen.getByText('Fund');
      expect(heading).toBeInTheDocument();
      const input = screen.getByLabelText('amount');
      await user.type(input, '200');
      const button = screen.getByRole('button', { name: /Fund/i });
      await user.click(button);
      const confirmBtn = await screen.findAllByRole('button', { name: /Fund/i });
      await user.click(screen.getByLabelText(/terms/i));
      await user.click(confirmBtn[1]);
      const modalContent = await screen.findByText(/Too Low/i);

      // ASSERT
      expect(modalContent).toBeInTheDocument();
      await user.click(screen.getByRole('button', { name: 'Close' }));
      expect(modalContent).not.toBeInTheDocument();
      const nullish = [...screen.queryAllByRole(/null/), ...screen.queryAllByRole(/undefined/)];
      expect(nullish).toHaveLength(0);
    });

    it('should render list items', async () => {
      const user = userEvent.setup();
      // ARRANGE
      render(
        <FundProvider bounty={bounty} refreshBounty={refreshBounty}>
          <FundPage bounty={bounty} />
        </FundProvider>
      );
      const input = screen.getByLabelText('amount');
      await user.type(input, '200');

      // ACT
      const token = screen.getByText('MATIC');
      const button = screen.getByRole('button', { name: /Fund/i });
      await user.click(button);
      const confirmBtn = await screen.findAllByRole('button', { name: /Fund/i });
      expect(confirmBtn[0]).toBeInTheDocument();
      await user.click(screen.getByLabelText(/terms/i));
      await user.click(confirmBtn[1]);
      const modalContent = await screen.findByText(/Too Low/i);

      // ASSERT
      expect(modalContent).toBeInTheDocument();

      // ASSERT
      expect(token).toBeInTheDocument();
      const nullish = [...screen.queryAllByRole(/null/), ...screen.queryAllByRole(/undefined/)];
      expect(nullish).toHaveLength(0);
    });

    it('should let user submit and handle too low amount of token', async () => {
      // ARRANGE
      const user = userEvent.setup();
      render(
        <FundProvider bounty={bounty} refreshBounty={refreshBounty}>
          <FundPage bounty={bounty} />
        </FundProvider>
      );
      // ACT
      const input = screen.getByLabelText('amount');
      await user.type(input, '200');
      const button = screen.getByRole('button', { name: /Fund/i });
      await user.click(button);
      const confirmBtn = await screen.findAllByRole('button', { name: /Fund/i });
      await user.click(screen.getByLabelText(/terms/i));
      await user.click(confirmBtn[1]);
      const modalContent = await screen.findByText(/Too Low/i);

      // ASSERT
      expect(modalContent).toBeInTheDocument();
      await user.click(screen.getByRole('button', { name: 'Close' }));
      expect(modalContent).not.toBeInTheDocument();
      const nullish = [...screen.queryAllByRole(/null/), ...screen.queryAllByRole(/undefined/)];
      expect(nullish).toHaveLength(0);
    });

    it('should let user submit and handle owned amount of Matic', async () => {
      // ARRANGE
      const user = userEvent.setup();
      const state = {
        ...InitialFundState,
        bounty,
        refreshBounty,
        token: Constants.tokenMetadata,
        volume: '3',
        approveTransferState: CONFIRM,
      };
      const dispatch = jest.fn();
      render(
        <FundContext.Provider value={[state, dispatch]}>
          <FundPage bounty={bounty} />
        </FundContext.Provider>
      );
      // ACT
      const value = await screen.findByText(/3 Dai/i);

      // ASSERT
      expect(value).toBeInTheDocument();
      const confirmBtn = await screen.findAllByRole('button', { name: /Fund/i });
      await user.click(confirmBtn[1]);
      expect(dispatch).toHaveBeenCalledWith({ type: 'SET_APPROVE_TRANSFER_STATE', payload: 'APPROVE' });
      const nullish = [...screen.queryAllByRole(/null/), ...screen.queryAllByRole(/undefined/)];
      expect(nullish).toHaveLength(0);
    });

    it('should let user submit and handle owned amount of Link', async () => {
      // ARRANGE
      const user = userEvent.setup();
      render(
        <FundProvider bounty={bounty} refreshBounty={refreshBounty}>
          <FundPage bounty={bounty} />
        </FundProvider>
      ); // ACT
      const input = screen.getByLabelText('amount');
      await user.type(input, '0.30sdf');
      await user.click(screen.getByText(/Matic/i));
      await user.click(screen.getAllByText(/link/i)[0]);
      const button = screen.getByRole('button', { name: /Fund/i });
      await user.click(button);
      const value = await screen.findByText(/.3 link/i);
      await user.click(screen.getByLabelText(/terms/i));

      // ASSERT
      expect(value);
      const confirmBtn = await screen.findByRole('button', { name: /Approv/i });
      await user.click(confirmBtn);
      const modalContent = await screen.findByText(/Transfer Complete/i, undefined, {
        timeout: 8000,
      });
      const tweet = screen.getByText('Tweet about it');
      expect(tweet).toBeInTheDocument();
      await user.click(screen.getByTestId('cross'));
      expect(modalContent).not.toBeInTheDocument();
      const nullish = [...screen.queryAllByRole(/null/), ...screen.queryAllByRole(/undefined/)];
      expect(nullish).toHaveLength(0);
    });

    it('should go straight to fund when DERC20 approved previously', async () => {
      // ARRANGE
      const user = userEvent.setup();
      render(
        <FundProvider bounty={bounty} refreshBounty={refreshBounty}>
          <FundPage bounty={bounty} />
        </FundProvider>
      );
      // ACT
      const input = screen.getByLabelText('amount');
      await user.type(input, '4.8');
      await user.click(screen.getByText(/Matic/i));
      await user.click(screen.getAllByText(/derc20/i)[0]);
      const button = screen.getByRole('button', { name: /Fund/i });
      await user.click(button);
      const value = await screen.findByText(/4.8 derc20/i);

      // ASSERT
      expect(value).toBeInTheDocument();
      await user.click(await screen.findByLabelText(/terms/i));

      const confirmBtn = await screen.findAllByRole('button', { name: /Fund/i });
      await user.click(confirmBtn[1]);
      const modalContent = await screen.findByText(/Transfer Complete/i);
      const tweet = screen.getByText('Tweet about it');
      expect(tweet).toBeInTheDocument();
      await user.click(screen.getByTestId('cross'));
      expect(modalContent).not.toBeInTheDocument();
      const nullish = [...screen.queryAllByRole(/null/), ...screen.queryAllByRole(/undefined/)];
      expect(nullish).toHaveLength(0);
    });

    it('should handle approval errors', async () => {
      // ARRANGE
      InitialState.openQClient.shouldError = true;
      const user = userEvent.setup();
      render(
        <FundProvider bounty={bounty} refreshBounty={refreshBounty}>
          <FundPage bounty={bounty} />
        </FundProvider>
      );
      // ACT
      const input = screen.getByLabelText('amount');
      await user.type(input, '0.30sdf');
      await user.click(await screen.findByText(/MATIC/i));
      await user.click(screen.getAllByText(/Chainlink/i)[0]);
      const button = await screen.findByRole('button', { name: /Fund/i });
      await user.click(button);

      // ASSERT
      expect(await screen.findByText(/0.3 LINK/));
      await user.click(await screen.findByLabelText(/terms/i));
      await user.click(await screen.findByRole('button', { name: /Approv/i }));
      const modalContent = await screen.findByText(/try again./i);
      expect(modalContent).toBeInTheDocument();

      await user.click(await screen.findByRole('button', { name: 'Close' }));
      expect(modalContent).not.toBeInTheDocument();
      const nullish = [...screen.queryAllByRole(/null/), ...screen.queryAllByRole(/undefined/)];
      expect(nullish).toHaveLength(0);
    });

    it('should prevent user from submitting over 10000000', async () => {
      // ARRANGE
      const user = userEvent.setup();
      render(
        <FundProvider bounty={bounty} refreshBounty={refreshBounty}>
          <FundPage bounty={bounty} />
        </FundProvider>
      );
      // ACT
      const input = screen.getByLabelText('amount');
      await user.type(input, '10000000');
      const button = await screen.findByRole('button', { name: /Fund/i });

      // ASSERT
      expect(button).toBeDisabled();
      await user.hover(button);
      const tooltip = await screen.findByText(/Must be between/);
      expect(tooltip).toBeInTheDocument();
      const nullish = [...screen.queryAllByRole(/null/), ...screen.queryAllByRole(/undefined/)];
      expect(nullish).toHaveLength(0);
    });

    it('should prevent user from submitting under 0.00000001', async () => {
      // ARRANGE
      const user = userEvent.setup();
      render(
        <FundProvider bounty={bounty} refreshBounty={refreshBounty}>
          <FundPage bounty={bounty} />
        </FundProvider>
      );
      // ACT
      const input = screen.getByLabelText('amount');
      await user.type(input, '0.00000001');
      const button = await screen.findByRole('button', { name: /Fund/i });

      // ASSERT
      expect(button).toBeDisabled();
      await user.hover(button);
      const tooltip = await screen.findByText(/Must be between/);
      expect(tooltip).toBeInTheDocument();
      const nullish = [...screen.queryAllByRole(/null/), ...screen.queryAllByRole(/undefined/)];
      expect(nullish).toHaveLength(0);
    });

    it('should show tooltip', async () => {
      // ARRANGE
      const user = userEvent.setup();
      render(
        <FundProvider bounty={bounty} refreshBounty={refreshBounty}>
          <FundPage bounty={bounty} />
        </FundProvider>
      ); // ACT / ASSERT
      const button = screen.getByRole('button', { name: /Fund/i });
      expect(button).toBeDisabled();
      await user.hover(button);
      const tooltip = await screen.findByText(/indicate the volume you'd like to fund with./i);
      expect(tooltip).toBeInTheDocument();
      const nullish = [...screen.queryAllByRole(/null/), ...screen.queryAllByRole(/undefined/)];
      expect(nullish).toHaveLength(0);
    });
  });
});
