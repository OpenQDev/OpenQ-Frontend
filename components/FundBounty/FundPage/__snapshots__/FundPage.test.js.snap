// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FundPage Split Price Bounties should lock token choice to deposit token when there is 1+ deposit 1`] = `
<DocumentFragment>
  <div
    class="flex-1 pt-4 pb-8 w-full max-w-[1200px] justify-center"
  >
    <div
      class="flex flex-col w-full space-y-5 pb-8 items-center md:border rounded-sm border-gray-700"
    >
      <div
        class="flex text-3xl w-full text-primary justify-center px-16 py-4 md:bg-[#161b22] md:border-b border-gray-700 rounded-t-sm"
      >
        Escrow Funds in Split Price Contract
      </div>
      <div
        class="flex flex-col space-y-5 w-5/6 pt-2"
      >
        <div
          class="flex w-full gap-4"
        >
          <div
            class="flex space-x-4 w-full undefined"
          >
            <div
              class="w-full flex w-full flex-row justify-between items-center px-4 input-field-big"
            >
              <div
                class="w-full bg-dark-mode"
              >
                <input
                  aria-label="amount"
                  autocomplete="off"
                  class="font-semibold number outline-none bg-input-bg text-primary w-full"
                  id="amount"
                  placeholder="0.0"
                  value=""
                />
              </div>
            </div>
            <div
              class="flex"
            >
              <button
                aria-label="select token"
                class="flex flex-row items-center p-0.5 px-2 cursor-default"
                disabled=""
              >
                 
                <div
                  class="flex h-4 w-4 items-center justify-center"
                >
                  <img
                    alt="selected token"
                    class="rounded-full"
                    data-nimg="1"
                    decoding="async"
                    height="40"
                    loading="lazy"
                    src="/crypto-logos/LINK.svg"
                    style="color: transparent;"
                    width="40"
                  />
                </div>
                <div
                  class="flex pl-2 pr-1 text-primary"
                >
                  LINK
                </div>
              </button>
            </div>
          </div>
        </div>
        <div
          class="flex gap-4"
        >
          <div
            class="w-full"
          >
            <div
              class="flex w-full input-field-big mb-4"
            >
              <div
                class=" flex items-center gap-3 w-full text-primary md:whitespace-nowrap"
              >
                <div
                  class="group "
                >
                  <div
                    class="cursor-help rounded-full border border-[#c9d1d9] aspect-square leading-4 h-4 box-content text-center font-bold text-primary"
                  >
                    ?
                  </div>
                  <div
                    class="justify-center w-full relative hidden z-50 group-hover:block  -top-1 "
                  >
                    <div
                      class="flex flex-col items-center"
                    >
                      <div
                        class="flex mt-0.5 md:mt-1 tooltip-triangle absolute undefined"
                      />
                      <div
                        class="flex tooltip absolute md:-left-12"
                      >
                        <div
                          class="whitespace-normal md:w-96 sm:w-60 w-40  "
                        >
                          This is the number of days that your deposit will be in escrow. After this many days, your deposit will be fully refundable if the contract has still not been claimed.
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <span>
                  Deposit Locked Period
                </span>
              </div>
              <div
                class="flex px-4 font-bold bg-input-bg"
              >
                <input
                  autocomplete="off"
                  class="text-primary text-right number outline-none w-full flex-1 bg-input-bg"
                  id="deposit-period"
                  value="30"
                />
              </div>
            </div>
            <div
              class="flex items-center"
            />
            <div
              class="group w-min"
            >
              <button
                class="text-center px-8 w-min items-center py-0.5  btn-default w-full cursor-not-allowed py-1.5"
                disabled=""
                type="button"
              >
                <div
                  class="text-center whitespace-nowrap w-full"
                >
                  Fund
                </div>
              </button>
              <div
                class="justify-center w-full relative hidden z-50 group-hover:block  -top-1  "
              >
                <div
                  class="flex flex-col items-center"
                >
                  <div
                    class="flex mt-0.5 md:mt-1 tooltip-triangle absolute undefined"
                  />
                  <div
                    class="flex tooltip absolute left-0"
                  >
                    <div
                      class="sm:w-40 md:w-60 whitespace-normal"
                    >
                      Please indicate the volume you'd like to fund with. Must be between 0.0000001 and 1,000,000.
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`FundPage Split Price Bounties should lock token choice to deposit token when there is 1+ deposit: Snapshot: button style changes to not appear as button 1`] = `
<button
  aria-label="select token"
  class="flex flex-row items-center p-0.5 px-2 cursor-default"
  disabled=""
>
   
  <div
    class="flex h-4 w-4 items-center justify-center"
  >
    <img
      alt="selected token"
      class="rounded-full"
      data-nimg="1"
      decoding="async"
      height="40"
      loading="lazy"
      src="/crypto-logos/LINK.svg"
      style="color: transparent;"
      width="40"
    />
  </div>
  <div
    class="flex pl-2 pr-1 text-primary"
  >
    LINK
  </div>
</button>
`;
